File: flutter/ios/Classes/VibeGrowthSdkPlugin.swift
import Flutter
import Foundation

public class VibeGrowthSdkPlugin: NSObject, FlutterPlugin {
    private var isInitialized = false
    private let exampleBaseUrlKey = "example.last_base_url"

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.vibegrowth.sdk/channel", binaryMessenger: registrar.messenger())
        let instance = VibeGrowthSdkPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]

        switch call.method {
        case "initialize":
            guard let appId = args?["appId"] as? String,
                  let apiKey = args?["apiKey"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "appId and apiKey are required", details: nil))
                return
            }
            let baseUrl = args?["baseUrl"] as? String
            VibeGrowthSDK.shared.initialize(appId: appId, apiKey: apiKey, baseUrl: baseUrl) { [weak self] success, error in
                if success {
                    self?.isInitialized = true
                    result(nil)
                } else {
                    result(FlutterError(code: "INIT_ERROR", message: error, details: nil))
                }
            }

        case "setUserId":
            guard ensureInitialized(result) else { return }
            guard let userId = args?["userId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "userId is required", details: nil))
                return
            }
            VibeGrowthSDK.shared.setUserId(userId)
            result(nil)

        case "getUserId":
            guard ensureInitialized(result) else { return }
            let userId = VibeGrowthSDK.shared.getUserId()
            result(userId)

        case "trackPurchase":
            guard ensureInitialized(result) else { return }
            let pricePaid = (args?["pricePaid"] as? Double) ?? (args?["amount"] as? Double)
            guard let pricePaid,
                  let currency = args?["currency"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "pricePaid and currency are required", details: nil))
                return
            }
            let productId = args?["productId"] as? String
            VibeGrowthSDK.shared.trackPurchase(pricePaid: pricePaid, currency: currency, productId: productId)
            result(nil)

        case "trackAdRevenue":
            guard ensureInitialized(result) else { return }
            guard let source = args?["source"] as? String,
                  let revenue = args?["revenue"] as? Double,
                  let currency = args?["currency"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "source, revenue, and currency are required", details: nil))
                return
            }
            VibeGrowthSDK.shared.trackAdRevenue(source: source, revenue: revenue, currency: currency)
            result(nil)

        case "trackSessionStart", "trackSession":
            guard ensureInitialized(result) else { return }
            guard let sessionStart = args?["sessionStart"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "sessionStart is required", details: nil))
                return
            }
            VibeGrowthSDK.shared.trackSessionStart(sessionStart: sessionStart)
            result(nil)

        case "getConfig":
            guard ensureInitialized(result) else { return }
            VibeGrowthSDK.shared.getConfig { configJson, error in
                if let error {
                    result(FlutterError(code: "CONFIG_ERROR", message: error, details: nil))
                } else {
                    result(configJson)
                }
            }

        case "setExampleBaseUrl":
            let baseUrl = (args?["baseUrl"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            let store = UserDefaultsStore()
            store.putString(exampleBaseUrlKey, value: baseUrl)
            result(nil)

        case "getExampleBaseUrl":
            let store = UserDefaultsStore()
            result(store.getString(exampleBaseUrlKey))

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func ensureInitialized(_ result: FlutterResult) -> Bool {
        guard isInitialized else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "VibeGrowthSDK is not initialized. Call initialize() first.", details: nil))
            return false
        }
        return true
    }
}

File: flutter/android/src/main/kotlin/com/vibegrowth/sdk/flutter/VibeGrowthSdkPlugin.kt
package com.vibegrowth.sdk.flutter

import android.content.Context
import com.vibegrowth.sdk.VibeGrowthSDK
import com.vibegrowth.sdk.persistence.PreferencesStore
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class VibeGrowthSdkPlugin : FlutterPlugin, MethodCallHandler {
    private val exampleBaseUrlKey = "example.last_base_url"

    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "com.vibegrowth.sdk/channel")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> {
                val appId = call.argument<String>("appId")
                val apiKey = call.argument<String>("apiKey")
                val baseUrl = call.argument<String>("baseUrl")
                if (appId == null || apiKey == null) {
                    result.error("INVALID_ARGS", "appId and apiKey are required", null)
                    return
                }
                VibeGrowthSDK.initialize(context, appId, apiKey, baseUrl, object : VibeGrowthSDK.InitCallback {
                    override fun onSuccess() {
                        result.success(null)
                    }

                    override fun onError(error: String) {
                        result.error("INIT_ERROR", error, null)
                    }
                })
            }
            "setUserId" -> {
                val userId = call.argument<String>("userId")
                if (userId == null) {
                    result.error("INVALID_ARGS", "userId is required", null)
                    return
                }
                try {
                    VibeGrowthSDK.setUserId(userId)
                    result.success(null)
                } catch (e: IllegalStateException) {
                    result.error("NOT_INITIALIZED", e.message, null)
                }
            }
            "getUserId" -> {
                try {
                    val userId = VibeGrowthSDK.getUserId()
                    result.success(userId)
                } catch (e: IllegalStateException) {
                    result.error("NOT_INITIALIZED", e.message, null)
                }
            }
            "trackPurchase" -> {
                val pricePaid = call.argument<Double>("pricePaid") ?: call.argument<Double>("amount")
                val currency = call.argument<String>("currency")
                val productId = call.argument<String>("productId")
                if (pricePaid == null || currency == null) {
                    result.error("INVALID_ARGS", "pricePaid and currency are required", null)
                    return
                }
                try {
                    VibeGrowthSDK.trackPurchase(pricePaid, currency, productId)
                    result.success(null)
                } catch (e: IllegalStateException) {
                    result.error("NOT_INITIALIZED", e.message, null)
                }
            }
            "trackAdRevenue" -> {
                val source = call.argument<String>("source")
                val revenue = call.argument<Double>("revenue")
                val currency = call.argument<String>("currency")
                if (source == null || revenue == null || currency == null) {
                    result.error("INVALID_ARGS", "source, revenue, and currency are required", null)
                    return
                }
                try {
                    VibeGrowthSDK.trackAdRevenue(source, revenue, currency)
                    result.success(null)
                } catch (e: IllegalStateException) {
                    result.error("NOT_INITIALIZED", e.message, null)
                }
            }
            "trackSessionStart", "trackSession" -> {
                val sessionStart = call.argument<String>("sessionStart")
                if (sessionStart == null) {
                    result.error("INVALID_ARGS", "sessionStart is required", null)
                    return
                }
                try {
                    VibeGrowthSDK.trackSessionStart(sessionStart)
                    result.success(null)
                } catch (e: IllegalStateException) {
                    result.error("NOT_INITIALIZED", e.message, null)
                }
            }
            "getConfig" -> {
                try {
                    VibeGrowthSDK.getConfig(object : VibeGrowthSDK.ConfigCallback {
                        override fun onSuccess(configJson: String) {
                            result.success(configJson)
                        }

                        override fun onError(error: String) {
                            result.error("CONFIG_ERROR", error, null)
                        }
                    })
                } catch (e: IllegalStateException) {
                    result.error("NOT_INITIALIZED", e.message, null)
                }
            }
            "setExampleBaseUrl" -> {
                val baseUrl = call.argument<String>("baseUrl")?.trim() ?: ""
                PreferencesStore(context).putString(exampleBaseUrlKey, baseUrl)
                result.success(null)
            }
            "getExampleBaseUrl" -> {
                result.success(PreferencesStore(context).getString(exampleBaseUrlKey))
            }
            else -> result.notImplemented()
        }
    }
}
